### ISSUE TITLE
Consider env vars when unmarshalling

### ISSUE THREAD

**Issue by aeneasr** [spf13/viper#188]
Is this possible?

**UPDATE: yes it is, see https://github.com/spf13/viper/issues/188#issuecomment-255519149**

``` go

type Config struct {
    BindPort int `mapstructure:"port" yaml:"port,omitempty"`
}

var c Config

// ...
viper.AutomaticEnv()

if err := viper.ReadInConfig(); err != nil {
    // ...
}

if err := viper.Unmarshal(&c); err != nil {
    // ...
}

// I want c.BindPort == viper.Get("PORT") == os.Getenv("PORT")
```

**Comment by rybit** [comment]
I am having the same problem, it isn't fixed directly by the #195. You have to register defaults it seems before viper will load the values. I have been experimenting with some terrible reflection code to go set defaults from a given structure, but I'd like to keep reflection out where possible. 

Consider the following code:
``` go

type Config struct {
  CAFiles  []string `mapstructure:"ca_files"
  CertFile string   `mapstructure:"cert_file"
  KeyFile  string   `mapstructure:"key_file"
  Servers  []string `mapstructure:"servers"`
}


func configuredViperInstance() *viper.Viper {
  v := viper.New()
  v.SetConfigName("config")
  v.AddConfigPath("./")

  v.SetEnvPrefix("test")
  v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
  v.AutomaticEnv()

  if err := v.ReadInConfig(); err != nil && !os.IsNotExist(err) {
    panic(err)
  }
  return v
}

func main() {
  noDefaults := configuredViperInstance()
  config := new(Config)
  if err := noDefaults.Unmarshal(config); err != nil {
    panic(err)
  }

  fmt.Printf("%+v\n", config)
  fmt.Printf("config-cert_file: %s\n", config.CertFile)
  fmt.Printf("cert_file: %s\n", noDefaults.GetString("cert_file"))

  fmt.Println("-------------------------")

  withDefaults := configuredViperInstance()
  withDefaults.SetDefault("cert_file", "")
  config = new(Config)
  if err := withDefaults.Unmarshal(config); err != nil {
    panic(err)
  }
  fmt.Printf("%+v\n", config)
  fmt.Printf("config-cert_file: %s\n", config.CertFile)
  fmt.Printf("cert_file: %s\n", noDefaults.GetString("cert_file"))
}
```

I then call it passing in this json 

``` json
{
  "servers": ["server-1", "server-2"],
  "key_file": "key-file"
}
```

``` sh
TEST_CERT_FILE="env-certfile" go run nested_conf_testing.go
```

I get the response 

``` sh
&{CAFiles:[] CertFile: KeyFile:key-file Servers:[server-1 server-2]}
config-cert_file:
cert_file: env-certfile
-------------------------
&{CAFiles:[] CertFile:env-certfile KeyFile:key-file Servers:[server-1 server-2]}
config-cert_file: env-certfile
cert_file: env-certfile
```

**Comment by arekkas** [comment]
@rybit you need to do

```go
type Config struct {
    BindPort int `mapstructure:"PORT" yaml:"port,omitempty"`
}

var c Config

// ...
viper.AutomaticEnv()

viper.BindEnv("PORT")
viper.SetDefault("PORT", 4444)

if err := viper.ReadInConfig(); err != nil {
    // ...
}

if err := viper.Unmarshal(&c); err != nil {
    // ...
}
```

to get this working.

**Comment by arekkas** [comment]
If BindEnv or SetDefault is missing, it won't work with the latest beggining with PR #195

**Comment by benoitmasson** [comment]
@rybit I don't think this is a bug, I'd say that this is the expected behavior: how should Viper know that a struct field has to be mapped to an environment variable, if you don't tell it to?

If you ask for config value `foo` with `viper.Get()`, Viper knows it has to look for a variable `FOO` in the environment and `viper.AutomaticEnv()` will be sufficient to handle that. But on the other side, `viper.AllSettings()` (used by `viper.Unmarshal()`) can't be aware that a config value for `foo` is expected (unless all environment values are added to the map, which wouldn't be reasonable).

As for me, automatically adding defaults or environment binding with reflection is the way to go, if done carefully. You could then use structure tags to have the programmer define his own default value, which is very handy and flexible (I do it myself :-) )

@arekkas From what I remember, if `AutomaticEnv` is on, then `SetDefault()` is enough (this tells Viper that such a configuration value exists). `BindEnv` is useful only when `AutomaticEnv` is off, and in that case, `SetDefault()` shouldn't be needed.
Are you sure both are necessary?

**Comment by burdiyan** [comment]
I wonder if it's possible to make @arekkas' example working without explicitly doing `BindEnv` and just doing `AutomaticEnv`.

Normally we use the following workflow in our applications:

1. Define global configuration structure as a Go struct.
2. Expose flags that can modify this configuration and bind the flag set to viper.
3. Allow to modify the configuration via file, flags or env variables.

Inside the application code we only deal with the config struct (to avoid dealing with arbitrary string keys in Viper). And we explicitly pass the config, it is not a global structure.

Basically the default values of the config are defined in flags. So I would like to avoid explicitly doing `BindEnv` for each of the env vars. It seems like it should be doable to have this with AutomaticEnv directly.

**Comment by burdiyan** [comment]
Oh, it seems like doing `viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))` makes it possible.

**Comment by from-nibly** [comment]
@burdiyan  were you able to confirm that your solution with the SetEnvKeyReplacer works? I am unable to get it to work in my project.  Maybe there is another thing I am missing?

calling `viper.GetString("foo.bar")` gets me the value but calling `viper.UnmarshalKey("foo", &cfg)` uses the value found in my config file.

**Comment by aarondl** [comment]
Odd. I do the same thing on the latest master branch and I still wind up with nothing being loaded into my struct. There's nowhere you actually define defaults? Or maybe your config file has empty values in it? When I try this `AllKeys` is totally empty unless I put things in the config file.

**Comment by from-nibly** [comment]
oh yes sorry just looked at it again. I define each property as "empty string" in my config file (which gives me the list of property names in `AllKeys`)

**Comment by krak3n** [comment]
I've also run into this problem so I wrote a quickie (and perhaps dirty) recursive function which manually calls `BindEnv` on a `struct` so you can have a partially complete config file:

``` go
func BindEnvs(iface interface{}, parts ...string) {
	ifv := reflect.ValueOf(iface)
	ift := reflect.TypeOf(iface)
	for i := 0; i < ift.NumField(); i++ {
		v := ifv.Field(i)
		t := ift.Field(i)
		tv, ok := t.Tag.Lookup("mapstructure")
		if !ok {
			continue
		}
		switch v.Kind() {
		case reflect.Struct:
			BindEnvs(v.Interface(), append(parts, tv)...)
		default:
			viper.BindEnv(strings.Join(append(parts, tv), "."))
		}
	}
}
```

Usage:

**Comment by natefinch** [comment]
@krak3n - thanks for the hack to get this to work.  I was considering doing the same thing, and happy to have someone else work out the fiddly bits.

For reference, mine defaults to use the lowercase name of the field instead of requiring the mapstructure tag (since almost all of mine are just straight strings):

```

func bindEnvs(v *viper.Viper, iface interface{}, parts ...string) {
	ifv := reflect.ValueOf(iface)
	ift := reflect.TypeOf(iface)
	for i := 0; i < ift.NumField(); i++ {
		fieldv := ifv.Field(i)
		t := ift.Field(i)
		name := strings.ToLower(t.Name)
		tag, ok := t.Tag.Lookup("mapstructure")
		if ok {
			name = tag
		}
		path := append(parts, name)
		switch fieldv.Kind() {
		case reflect.Struct:
			bindEnvs(v, fieldv.Interface(), path...)
		default:
			v.BindEnv(strings.Join(path, "."))
		}
	}
}
```

**Comment by supershal** [comment]
all the options suggested above either required to set config file with empty values or bind each environment variables. These options are prone to errors.

I tried following simple workaround based on reading comments this issue. This does not require to bind env values or set defaults or set file with all config set to `EMPTY VALUE` or reflection.

```
type Config struct {
	Name string `yaml:"name"`
	Port int `yaml:"port"`	
}
func NewConfig() *Config {
	return &Config{Port: 8080, Name: "foo"}
}

var cfgFile = "/tmp/overwrite.yaml" // contents: name: bar
var config *Config
func initConfigE() error {
	v := viper.New()
	// set default values in viper.
	// Viper needs to know if a key exists in order to override it.
	// https://github.com/spf13/viper/issues/188
	b, err := yaml.Marshal(NewConfig())
	if err != nil {
		return err
	}
	defaultConfig := bytes.NewReader(b)
	v.SetConfigType("yaml")
	if err := v.MergeConfig(defaultConfig); err != nil {
		return err
	}
	// overwrite values from config
	v.SetConfigFile(cfgFile)
	if err := v.MergeInConfig(); err != nil {
		if _, ok := err.(viper.ConfigParseError); ok {
			return err
		}
		// dont return error if file is missing. overwrite file is optional
	}
	// tell viper to overrwire env variables
	v.AutomaticEnv()
	v.SetEnvPrefix(envVarPrefix)
	v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))

	// refresh configuration with all merged values
	config = &Config{}
	return v.Unmarshal(&config)
}
```

**Comment by krak3n** [comment]
@supershal nice approach, nice way to avoid reflection. I'll see if this can work with `toml` formatted configs tool :+1:

**Comment by unmultimedio** [comment]
Hello, any updates on this? `getKeys` func still does not read from ENVwhen no config file or defults are set.

**Comment by sagikazarmark** [comment]
Please see and follow the issue above. ☝️

**Comment by segevfiner** [comment]
If `mapstructure` either returned metadata about fields not unmarshalled into in the target struct (Rather then the source), or otherwise had a callback to handle such fields, it could have been an easy way for viper to fix this. Otherwise using reflection in a similar way to what `mapstructure` does to handle the struct seems like the most viable approach, though it will have to correctly handle the various `mapstructure` tags. You can find various snippets of code doing just that around the issue tracker and the internet. Even then, it still won't work with dynamic parts of the struct, e.g. `map[string]interface{}` and so on, without further cleverness from viper rather then it just doing something akin to `BindEnv`.

Serializing the struct won't work correctly with `omitempty`.

**Comment by 89z** [comment]
https://github.com/spf13/viper/issues/140

https://github.com/spf13/viper/issues/188
https://github.com/spf13/viper/issues/212
https://github.com/spf13/viper/issues/522
https://github.com/spf13/viper/issues/576
https://github.com/spf13/viper/issues/761
https://github.com/spf13/viper/issues/911

**Comment by tunhvn** [comment]
I have the same issue but with flags. Unmarshal() works only with value from config file, it doesn't check flags.

**Comment by Kavit900** [comment]
Hi Viper team, so I have set environment variables in my docker container and I am trying to use viper to get those values and then use unmarshall but it doesn't seem to work, does viper only works if I have a separate environment file?

**Comment by 76creates** [comment]
> all the options suggested above either required to set config file with empty values or bind each environment variables. These options are prone to errors.
> 
> I tried following simple workaround based on comments from this issue. It does not require to bind env values or set defaults or set file with all config set to `EMPTY VALUE` or reflection.
> 
> ```go
> type Config struct {
> 	Name string `yaml:"name"`
> 	Port int `yaml:"port"`	
> }
> func NewConfig() *Config {
> 	return &Config{Port: 8080, Name: "foo"}
> }
> 
> var cfgFile = "/tmp/overwrite.yaml" // contents: name: bar
> var config *Config
> func initConfigE() error {
> 	v := viper.New()
> 	// set default values in viper.
> 	// Viper needs to know if a key exists in order to override it.
> 	// https://github.com/spf13/viper/issues/188
> 	b, err := yaml.Marshal(NewConfig())
> 	if err != nil {
> 		return err
> 	}
> 	defaultConfig := bytes.NewReader(b)
> 	v.SetConfigType("yaml")
> 	if err := v.MergeConfig(defaultConfig); err != nil {
> 		return err
> 	}
> 	// overwrite values from config
> 	v.SetConfigFile(cfgFile)
> 	if err := v.MergeInConfig(); err != nil {
> 		if _, ok := err.(viper.ConfigParseError); ok {
> 			return err
> 		}
> 		// dont return error if file is missing. overwrite file is optional
> 	}
> 	// tell viper to overrwire env variables
> 	v.AutomaticEnv()
> 	v.SetEnvPrefix(envVarPrefix)
> 	v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
> 
> 	// refresh configuration with all merged values
> 	config = &Config{}
> 	return v.Unmarshal(&config)
> }
> ```

reading this I came to understanding that `SetEnvKeyReplacer` is doing complete opposite of what I taught it did, i was setting it to `strings.NewReplacer("_", ".")`, might be good idea to add the explanation/example to the function

**Comment by wsgomes** [comment]
This is my solution:
```go
type Configurations struct {
	Database DatabaseConfigs
}

type DatabaseConfigs struct {
	User string
	Pass string
	Host string
	Port int
	Name string
}

var configs Configurations

func init() {
	viper.SetConfigName("config")
	viper.AddConfigPath("config")
	viper.SetConfigType("yml")

	if err := viper.ReadInConfig(); err != nil {
		logger.Fatal("config: error reading config file: " + err.Error())
	}

	for _, key := range viper.AllKeys() {
		envKey := strings.ToUpper(strings.ReplaceAll(key, ".", "_"))
		err := viper.BindEnv(key, envKey)
		if err != nil {
			logger.Fatal("config: unable to bind env: " + err.Error())
		}
	}

	if err := viper.Unmarshal(&configs); err != nil {
		logger.Fatal("config: unable to decode into struct: " + err.Error())
	}
}
```
1. Do not use `viper.AutomaticEnv()`.
2. From viper key `database.host`, create envKey `DATABASE_HOST`.
3. BindEnv: key => envKey
